src/Language/PureScript/Ide/Imports.hs

```diff
@@ -0,0 +1,193 @@
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PackageImports        #-}
+
+module Language.PureScript.Ide.Imports
+       ( addImplicitImport
+       , addImportForIdentifier
+       , answerRequest
+         -- for tests
+       , addImplicitImport'
+       , addExplicitImport'
+       , sliceImportSection
+       , prettyPrintImport'
+       , Import(Import)
+       )
+       where
+
+import           Control.Monad.Error.Class
+import           Control.Monad.IO.Class
+import           "monad-logger" Control.Monad.Logger
+import           Data.Function (on)
+import qualified Data.List                          as List
+import           Data.Maybe                         (mapMaybe)
+import           Data.Monoid                        ((<>))
+import           Data.Text                          (Text)
+import qualified Data.Text                          as T
+import qualified Data.Text.IO                       as TIO
+import qualified Language.PureScript                as P
+import           Language.PureScript.Ide.Completion
+import           Language.PureScript.Ide.Error
+import           Language.PureScript.Ide.Filter
+import           Language.PureScript.Ide.State
+import           Language.PureScript.Ide.Types
+
+data Import = Import P.ModuleName P.ImportDeclarationType  (Maybe P.ModuleName)
+              deriving (Eq, Show)
+
+-- | Reads a file and returns the (lines before the imports, the imports, the
+-- lines after the imports)
+parseImportsFromFile :: (MonadIO m, MonadError PscIdeError m) =>
+                        FilePath -> m ([Text], [Import], [Text])
+parseImportsFromFile fp = do
+  file <- liftIO (TIO.readFile fp)
+  pure (sliceImportSection (T.lines file))
+
+sliceImportSection :: [Text] -> ([Text], [Import], [Text])
+sliceImportSection ls =
+  let
+    preImportSection = takeWhile (not . hasImportPrefix) ls
+    importSection =
+      takeWhile continuesImport $
+        dropWhile (not . hasImportPrefix) ls
+    postImportSection =
+      dropWhile continuesImport $
+        dropWhile (not . hasImportPrefix) ls
+    hasImportPrefix = T.isPrefixOf "import"
+    continuesImport x = hasImportPrefix x || T.isPrefixOf " " x || x == ""
+  in (preImportSection, parseImports importSection, postImportSection)
+
+
+-- | Concatenates multiline imports into a single line and tries to parse them.
+-- Anything that fails to parse gets left out
+parseImports :: [Text] -> [Import]
+parseImports ts =
+  let
+    concatMultilineImports = foldl step [] ts
+    step :: [Text] -> Text -> [Text]
+    step acc t = if T.isPrefixOf " " t
+                 then init acc ++ [last acc <> t]
+                 else acc ++ [t]
+  in
+    mapMaybe parseImport concatMultilineImports
+
+parseImport :: Text -> Maybe Import
+parseImport t =
+  case P.lex "<psc-ide>" (T.unpack t)
+       >>= P.runTokenParser "<psc-ide>" P.parseImportDeclaration' of
```

Comment:
Since you're parsing anyway, I think it might be best to just parse the whole module. It might be slightly slower, but shouldn't be too much slower, and it'll be less fragile IMO.